C++: Allocation.qll: Analyze common pattern of malloc() invocations to provide more accurate getSizeMult()

Cornelius-Riemenschneider opened the pull request:
AllocationExpr.getSizeMult() now analyzes the size expression of function calls.

This yields more precise size information in a lot of the common cases of C allocation code,
as the common pattern malloc(count * sizeof(type)) is now understood.

MathiasVP commented on cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll:
```diff
@@ -255,6 +255,24 @@ class OperatorNewAllocationFunction extends AllocationFunction {
   }
 }
 
+pragma[inline]
+private predicate deconstructSizeExpr(Expr sizeExpr, Expr lengthExpr, int sizeof) {
```
Why `pragma[inline]`?

Cornelius-Riemenschneider replied:
The idea is to not compute the predicate on all Exprs (the second case already includes almost all expressions), but only for those expressions that are the size argument to an AllocationExpr.
If you don't like this, I can remove it.

geoffw0 replied:
We generally prefer to let the optimizer make decisions like this unless the result is demonstrably performing worse than it should be.  This keeps the QL clean and easy to maintain.

That said, if removing `pragma[inline]` drastically increases the time spend computing this (or a related) predicate, then lets keep it.

Cornelius-Riemenschneider replied:
:+1:

Cornelius-Riemenschneider commented on cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll:
```diff
@@ -272,15 +290,29 @@ class CallAllocationExpr extends AllocationExpr, FunctionCall {
+        )
+        or
+        not exists(Expr lengthExpr | deconstructSizeExpr(sizeExpr, lengthExpr, _)) and
+        result = sizeExpr
```
Good catch, that's an artifact of a previous iteration -> removed.

Cornelius-Riemenschneider commented:
Fixed a bug (see commit message) and addressed the review.

The big picture of this is that I want to use the information from here (and #3337) to power the array length tracking library I am writing.
Thus, having array lengths and sizes available for (most) malloc allocations is crucial.

You can have a look at the preview here:
https://github.com/Cornelius-Riemenschneider/ql/pull/1
And you can see how much code I can get rid of by using these new predicates from the allocation library:
https://github.com/Cornelius-Riemenschneider/ql/pull/1/commits/0b256d0f7e0cce62ffe0bbf703cdd80d848174da

I hope this makes sense, if not I'm happy to elaborate further.

jbj commented on cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll:
```diff
@@ -255,6 +255,36 @@ class OperatorNewAllocationFunction extends AllocationFunction {
+      lengthExpr = sizeExpr.(MulExpr).getAnOperand() and
+      not lengthExpr instanceof SizeofOperator and
+      sizeof = sizeofOp.getValue().toInt()
+    )
```
Why is it important that we're multiplying by a `sizeof` rather than a constant value (anything with a `getValue` result) that was obtained by other means? (This is not a question that needs to block this PR).

Cornelius-Riemenschneider replied:
Two reasons:
1. Using `sizeof`, the user clearly states their intent as using a size multiplier (the predicate is named `getSizeMult`, so it shouldn't return values that are not clearly sizes of the allocation.
This could go even further by using the `getAllocatedElementType` from #3337, and ensuring that either `getSizeMult()=1` or ´getSizeMult() = getAllocatedElementType().getSize() = getSizeMult` holds.
2. I don't think  we want that the predicate has multiple results, so in the cases `malloc(5*sizeof(int))` and `malloc(5*3)` we'd need a canonical way to only report one size multiplier (else predicates like `getSizeBytes` would also evaluate to multiple values)

jbj replied:
That sounds reasonable. We can always adapt to coding patterns we observe later.

Cornelius-Riemenschneider commented:
A counterpoint to merging this pull request would be that, should we wish to extend support of understanding allocation expressions like `malloc(sizeof(somestruct)+lenOfBuffer)` (i.e. structs with a variable-length buffer at the end), this API is not suited for generalization.

jbj commented:
> A counterpoint to merging this pull request would be that, should we wish to extend support of understanding allocation expressions like `malloc(sizeof(somestruct)+lenOfBuffer)` (i.e. structs with a variable-length buffer at the end), this API is not suited for generalization.

The API is already there, so I don't think it saves us anything to avoid merging this PR. If we want to support _flexible array members_, we can make a separate API for that.

Cornelius-Riemenschneider commented:
Can somebody tell me which tests are failing? I can't access the Github-internal infrastructure.

jbj commented:
```
00:08:41  Some tests failed!
00:08:41      ql/cpp/ql/test/library-tests/allocators/allocators.ql: Test failure (Expected output does not match)
00:08:41      ql/cpp/ql/test/query-tests/Critical/OverflowCalculated/NoSpaceForZeroTerminator.qlref: Test failure (Expected output does not match)
00:08:41      ql/cpp/ql/test/query-tests/Critical/OverflowCalculated/OverflowCalculated.qlref: Test failure (Expected output does not match)
00:08:41      ql/cpp/ql/test/query-tests/Security/CWE/CWE-131/semmle/NoSpaceForZeroTerminator/NoSpaceForZeroTerminator.qlref: Test failure (Expected output does not match)
```

Cornelius-Riemenschneider commented:
The test failures are in two categories:
allocators.ql was not accepted previously by mistakte (after the last bugfix).
The failing overflow queries now detect more real bugs, in cases where getSizeExpr() is now more precise (i.e. doesn't report the sizeof() multiplication, but the variable read of the length that is allocated).
All test output has been accepted, no code change necessary.
